core/commands/storage.go

```diff
@@ -197,45 +192,21 @@ var storageUploadInitCmd = &cmds.Command{
 		// TODO: Verify ch to get CHR
 
 		// RemoteCall(user, CHID, CHR) to get signedPayment
-		argRespc := []string{hash, r.ID, r.Challenge}
-		signedPaymentBody, err := remoteCall.CallGet("/storage/upload/respc", argRespc)
+		signedPaymentBody, err := p2pCall(n, pid, "/storage/upload/respc", hash, r.ID, r.Challenge)
 		if err != nil {
-			log.Error("fail to get resp with signedPayment: ", err)
 			return err
 		}
 		var halfSignedChannelState ledgerPb.SignedChannelState
 		err = proto.Unmarshal(signedPaymentBody, &halfSignedChannelState)
 		if err != nil {
-			log.Error("fail to unmarshal signed payment: ", err)
 			return err
 		}
-		channelState := halfSignedChannelState.GetChannel()
 
-		// Verify payment
-		pk, err := pid.ExtractPublicKey()
-		if err != nil {
-			return err
-		}
-		ok, err = ledger.Verify(pk, channelState, halfSignedChannelState.GetFromSignature())
-		if err != nil || !ok {
-			log.Error("fail to verify channel state: ", err)
-			return err
-		}
-
-		// sign with private key
-		selfPrivKey := n.PrivateKey
-		sig, err := ledger.Sign(selfPrivKey, channelState)
-		if err != nil {
-			log.Error("fail to sign: ", err)
-			return err
-		}
-		halfSignedChannelState.ToSignature = sig
-		SignedchannelState := halfSignedChannelState
+		SignedchannelState, err := verifyAndSign(pid, n, &halfSignedChannelState)
 
 		// Close channel
-		err = ledger.CloseChannel(req.Context, ledgerClient, &SignedchannelState)
+		err = ledger.CloseChannel(req.Context, ledgerClient, SignedchannelState)
```

Comment:
How about moving this into the `verifySign` and call it `verifySignClose`, since you need to open a connection

Reply:
Only close channel would require connection, sign can totally offline since it's only a preparation process